HTTP headers: deprecating "(Want-)?Digest", adding "(Want-)?(Content-|Repr-)Digest" (#26885)
mdn/content: files/en-us/web/http/headers/repr-digest/index.md

---
title: Repr-Digest
slug: Web/HTTP/Headers/Repr-Digest
page-type: http-header
browser-compat: http.headers.Repr-Digest
spec-urls: https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-digest-headers-12
status:
  - experimental
---

{{HTTPSidebar}}{{SeeCompatTable}}

> The **`Repr-Digest`** response or request header provides a {{Glossary("digest"}} of the [selected representation](https://www.rfc-editor.org/rfc/rfc9110#section-6.4) of the target resource. It is invariant under e.&hairsp;g. {{HTTPHeader("Content-Encoding")}} or {{HTTPHeader("Content-Range")}}, which do affect the {{HTTPHeader("Content-Digest")}}. Note furthermore that [Content Negotiation](/en-US/docs/Web/HTTP/Content_negotiation) can result in different selected representation with respectively different representation digests.

Reviewer: ```suggestion
The **`Repr-Digest`** response or request header provides a {{Glossary("digest"}} of the [selected representation](https://www.rfc-editor.org/rfc/rfc9110#section-6.4) of the target resource. It is invariant under e.g., {{HTTPHeader("Content-Encoding")}} or {{HTTPHeader("Content-Range")}}, which do affect the {{HTTPHeader("Content-Digest")}}. Furthermore, [Content Negotiation](/en-US/docs/Web/HTTP/Content_negotiation) can result in different selected representations with different representation digests.
```

<table class="properties">
  <tbody>
    <tr>
      <th scope="row">Header type</th>
      <td>{{Glossary("Representation header")}}</td>
    </tr>
    <tr>
      <th scope="row">{{Glossary("Forbidden header name")}}</th>
      <td>no</td>
    </tr>
  </tbody>
</table>

## Syntax

> `Repr-Digest` describes an [RFC8941 dictionary](https://www.rfc-editor.org/rfc/rfc8941#section-3.2) with its keys being names of digest algorithms and its values being the digest in bytes (or an integer digest for legacy digest algorithms).
> 
> > **Note:** In contrast to earlier drafts of the specification, the standard-base64-encoded digest bytes are wrapped in colons (`:`, ASCII 0x3A) as part of the [dictionary syntax](https://www.rfc-editor.org/rfc/rfc8941#name-byte-sequences).

Reviewer 2: Does not match the template https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Page_types/HTTP_header_page_template

Unless absolutely necessary, do not link the spec. It's our job to show how the API is used "for humans".

```suggestion

`Repr-Digest` directives dictionaries, with keys being names of digest algorithms and values being the digest in bytes, wrapped in colons (`:`, ASCII 0x3A).

```http-nolint
# Secure digest algorithms
Repr-Digest sha512=:calculated_digest:
Repr-Digest: sha-256=:calculated_digest:

# Insecure digest algorithms
Repr-Digest md5=:calculated_digest:
Repr-Digest sha=:calculated_digest:
Repr-Digest unixsum=:calculated_digest:
Repr-Digest unixcksum=:calculated_digest:
Repr-Digest adler=:calculated_digest:
Repr-Digest crc32c=:calculated_digest:
```

> 
(or an integer digest for legacy digest algorithms).
The 
> **Note:** Legacy digest algorithms (in earlier drafts of the specification), use an integer digest and do nor wrap it in colons.
```

Author: I think it is much more human-readable if one does not say "a dictionary" (which is very ambiguous) but rather "an RFC8941 dictionary" which specifically states which syntactical form of a dictionary is at play here.

Furthermore, half-correctly and vaguely describing the dictionary as you have done benefits no one.

Reviewer: I wonder if we should have a glossary entry for RFC8941 dictionary: I don't like to send readers to the spec (that are difficult to understand) to grasp the syntax.

## Directives

### Digest algorithms

Only two digest algorithms are currently registered and not considered insecure: `sha-512` and `sha-256`.

The registered insecure digest algorithms are: `md5`, `sha` (SHA-1), `unixsum`, `unixcksum`, `adler` (ADLER32) and `crc32c`.

Usage of digest algorithms which are considered insecure is discouraged as collisions can realistically be forced, rendering the digest's usefulness weak. Unless when working with legacy systems (which is unlikely since most will expect the legacy {{HTTPHeader("Digest")}} header and not understand this specification), consider not emitting a `Repr-Digest` instead of emitting one using an insecure digest algorithm.

## Examples

> ### Example: HTTP response where `Repr-Digest` and `Content-Digest` coincide

Reviewer: ```suggestion
### HTTP response where `Repr-Digest` and `Content-Digest` coincide
```

An HTTP server may send content octets equivalent to the selected representation's octets:

```http
...
Repr-Digest: sha-256=:AEGPTgUMw5e96wxZuDtpfm23RBU3nFwtgY5fw4NYORo=:
Content-Digest: sha-256=:AEGPTgUMw5e96wxZuDtpfm23RBU3nFwtgY5fw4NYORo=:
...
Content-Type: text/yaml
Content-Encoding: identity
Content-Length: 38054
Content-Range: 0-38053/38054
...
```

> ### Example: HTTP responses where {{HTTPHeader("Repr-Digest")}} and {{HTTPHeader("Content-Digest")}} diverge

Reviewer: ```suggestion
### HTTP responses where {{HTTPHeader("Repr-Digest")}} and {{HTTPHeader("Content-Digest")}} diverge
```

A static file server may however choose to compress an HTML page, resulting in differing {{HTTPHeader("Content-Digest")}} and `Repr-Digest` header values:

```http
...
Repr-Digest: sha-256=:AEGPTgUMw5e96wxZuDtpfm23RBU3nFwtgY5fw4NYORo=:, sha-512=:U59TCCaZPA9Qio3CzHJVAgDnIAut53t5Sgkj2Gv4BvDd0b+OX9QpIdgWkzdXLmBsmvBrf3t5PBt+UrVK6k5dkw==:
Content-Digest: sha-256=:293wcr5IoFAsDCzdoDXR1Qppgf2yxOPO1bvQ3nZQtuI=:, unixsum=54809
...
Content-Type: text/html; charset=utf-8
Content-Encoding: br
...

...
```

```http
...
Repr-Digest: sha-256=:AEGPTgUMw5e96wxZuDtpfm23RBU3nFwtgY5fw4NYORo=:, sha-512=:U59TCCaZPA9Qio3CzHJVAgDnIAut53t5Sgkj2Gv4BvDd0b+OX9QpIdgWkzdXLmBsmvBrf3t5PBt+UrVK6k5dkw==:
Content-Digest: sha-256=:rv9Jivc4TmcacLUshzN3OdX7Hz+ORnQRaiTaIKZQ0zk=:
...
Content-Type: text/html; charset=utf-8
Content-Encoding: deflate, deflate, deflate
...

...
```

> ### Example: Successful HTTP request-response employing {{HTTHeader("Want-Repr-Digest")}}, {{HTTHeader("Repr-Digest")}}, {{HTTHeader("Content-Digest")}}

Reviewer: ```suggestion
### Successful HTTP request-response employing {{HTTHeader("Want-Repr-Digest")}}, {{HTTHeader("Repr-Digest")}}, {{HTTHeader("Content-Digest")}}
```

```http
PUT /api/transact HTTP/1.1
Want-Repr-Digest: sha-256=8
Content-Type: text/json
...
```

```http
HTTP/1.1 201 Created
Repr-Digest: sha-256=:W8oN3H3CmE/CBpV6ZPNozV2AIDzzQpWL7CCOXyDyDzI=:
Content-Encoding: br
Content-Digest: sha-256=:2IBI7hQn83oTCgB3Z/6apOl91WGoctRfRj/F9gkvVo8=:
...
```

> ### Example: Unsuccessful HTTP request-response employing {{HTTHeader("Repr-Digest")}}

Reviewer: ```suggestion
### Unsuccessful HTTP request-response employing {{HTTHeader("Repr-Digest")}}
```

<table class="properties">
  <tbody>
    <tr>
      <th scope="row">Header type</th>
      <td>{{Glossary("Representation header")}}</td>
    </tr>
    <tr>
      <th scope="row">{{Glossary("Forbidden header name")}}</th>
      <td>no</td>
    </tr>
  </tbody>
</table>

## Syntax

```http
Repr-Digest <directive_dictionary>
```

## Directives

`Repr-Digest` directives (`<directive_dictionary>`) are dictionaries with keys being names of digest algorithms and values being the digest in bytes, wrapped in colons (`:`, ASCII 0x3A).
For example:

```http
Repr-Digest: sha-256=:AEGPTgUMw5e96wxZuDtpfm23RBU3nFwtgY5fw4NYORo=:
```

The registered algorithms that are considered secure are: `sha-512` and `sha-256`.

The registered algorithms that are considered insecure are: `md5`, `sha` (SHA-1), `unixsum`, `unixcksum`, `adler` (ADLER32) and `crc32c`.

> **Note:** Using insecure digest algorithms is discouraged, because collisions can realistically be forced (making the digests less useful).

> > It is often better to not to emit a `Repr-Digest` at all than to emit one that specifies an insecure digest algorithm.

Reviewer 2: FYI, I updated this to try put in format acceptable to the template. Let's see what @teoli2003 says.

Author: Your statement is incorrect: e. g. "unixcksum" is not a byte array dictionary value.

{{HTTPSidebar}}{{SeeCompatTable}}

> The **`Repr-Digest`** response or request header provides a {{Glossary("digest"}} of the [selected representation](https://www.rfc-editor.org/rfc/rfc9110#section-6.4) of the target resource. It is invariant under e.&hairsp;g. {{HTTPHeader("Content-Encoding")}} or {{HTTPHeader("Content-Range")}}, which do affect the {{HTTPHeader("Content-Digest")}}. Note furthermore that [Content Negotiation](/en-US/docs/Web/HTTP/Content_negotiation) can result in different selected representation with respectively different representation digests.

Reviewer 2: Remove all cases of `e.hairsp;`

Author: Did you mean to replace my subphrase?

Reviewer 2: No. 

What I am saying here is that I don't see the benefit of ``e.hairsp;g.` over `e.g.`

## Examples

> ### HTTP response where `Repr-Digest` and `Content-Digest` coincide

Author: Is markup not allowed in headers or is this edit of your own taste? Looking at <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers>, HTTP headers seem to be marked up as inline code.

Reviewer 3: Markup is allowed. tbh all that happened here was that I removed the macro calls from some of the headers, which also removed the code style, then I removed it here just to be consistent. I don't mind either way, and if you'd like to put the code style back (not the macros though) I am happy with that.

Author: Done in bf6ba95d087150a1b413dd1363b28069b9bc8bc8.
